### PR TITLE
Add RpcSystemLog handler

### PR DESCRIPTION
**Pull request by yojagad**
Part of work for #3117

### CODE REVIEW

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 1]
```diff
@@ -362,23 +365,33 @@ internal void InvokeResponse(InvocationResponse invokeResponse)
             }
         }
 
+        internal void SystemLog(RpcEvent msg)
+        {
+            var rpcLog = msg.Message.RpcSystemLog;
+            LogContents(rpcLog.InvocationId, $"SystemLog: {rpcLog.Message}", rpcLog.Exception.Message, rpcLog.Exception.StackTrace, null, (LogLevel)rpcLog.Level);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: System logs need to be logged using worker logger.

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 2]
```diff
@@ -362,23 +365,33 @@ internal void InvokeResponse(InvocationResponse invokeResponse)
...
                 // Restore the execution context from the original invocation. This allows AsyncLocal state to flow to loggers.
                 System.Threading.ExecutionContext.Run(context.AsyncExecutionContext, (s) =>
                 {
-                    if (rpcLog.Exception != null)
+                    if (exceptionMessage != null)
                     {
-                        var exception = new RpcException(rpcLog.Message, rpcLog.Exception.Message, rpcLog.Exception.StackTrace);
-                        context.Logger.Log(logLevel, new EventId(0, rpcLog.EventId), rpcLog.Message, exception, (state, exc) => state);
+                        var exception = new RpcException(message, exceptionMessage, stackTrace);
+                        context.Logger.Log(logLevel, new EventId(0, eventId), message, exception, (state, exc) => state);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: if there is an exception, lets log using Log.Error

Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: context.Logger.Log should be used only for invocation logs. Also, need to figure out if a language worker logs system log within the invocation context, is it possbile to populate invoation id, function name etc as part of the system log?

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 3]
```diff
@@ -80,9 +79,13 @@ internal class LanguageWorkerChannel : ILanguageWorkerChannel, IDisposable
                 .Where(msg => msg.WorkerId == _workerId);
 
             _eventSubscriptions.Add(_inboundWorkerEvents
-                .Where(msg => msg.MessageType == MsgType.RpcLog)
+                .Where(msg => msg.MessageType == MsgType.RpcLog && (msg.Message.RpcLog.Category == null || !msg.Message.RpcLog.Category.Equals("System", StringComparison.OrdinalIgnoreCase)))
                 .Subscribe(Log));
 
+            _eventSubscriptions.Add(_inboundWorkerEvents
+               .Where(msg => (msg.MessageType == MsgType.RpcLog) && (msg.Message.RpcLog.Category != null && msg.Message.RpcLog.Category.Equals("System", StringComparison.OrdinalIgnoreCase)))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: move this to separate helper method IsRpcSystemLog and reuse this method above as well. and add unit tests

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 4]
```diff
@@ -454,9 +488,9 @@ public void Dispose()
             Dispose(true);
         }
 
-        private bool IsTriggerMetadataPopulatedByWorker()
```
Comment (mhoeger) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: Why is this getting removed? maybe a merge conflict?

Comment (yojagad) [reply src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: oh no! yes. Thanks for catching that.

**Comment by yojagad** [comment]
> Please update proto file and update this PR

@pragnagopa Updated the PR.

File: src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs [hunk 5]
```diff
@@ -389,6 +393,36 @@ internal void Log(RpcEvent msg)
             }
         }
 
+        internal void SystemLog(RpcEvent msg)
+        {
+            RpcLog systemLog = msg.Message.RpcLog;
+            LogLevel logLevel = (LogLevel)systemLog.Level;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerChannel.cs]: this conversion might not be guaranteed. Add default switch to log as informational.

File: test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs [hunk 6]
```diff
@@ -108,6 +109,14 @@ public void SendWorkerInitRequest_PublishesOutboundEvent()
             Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, _expectedLogMsg)));
         }
 
+        [Fact]
+        public void SendSystemLogMessage_PublishesSystemLogMessage()
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs]: - add test for isRpcSystemLog
- add tests for different categories - warning and error as well
- add test where Rpc LogLevel does not convert to a valid LogLevle

File: test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs [hunk 7]
```diff
@@ -108,6 +109,18 @@ public void SendWorkerInitRequest_PublishesOutboundEvent()
             Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, _expectedLogMsg)));
         }
 
+        [Theory]
+        [InlineData(RpcLog.Types.Level.Information, RpcLog.Types.Level.Information)]
+        [InlineData(RpcLog.Types.Level.Error, RpcLog.Types.Level.Error)]
+        [InlineData(RpcLog.Types.Level.Warning, RpcLog.Types.Level.Warning)]
+        [InlineData(RpcLog.Types.Level.Trace, RpcLog.Types.Level.Information)]
+        public void SendSystemLogMessage_PublishesSystemLogMessage(RpcLog.Types.Level levelToTest, RpcLog.Types.Level expectedLogLevel)
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs]: what about invalid LogLevel? RpcLogLevel might not convert to a valid LogLevel. What happens then?

Comment (yojagad) [reply test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs]: Trace is an invalid LogLevel. Expected is Information. (That's the last InlineData)

**Pull request merged**